File: sim/build_benchmark_manifest.py
from __future__ import annotations

import argparse
import csv
import json
from pathlib import Path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Build a SLURM manifest for BMEDIATOR simulation benchmarks.")
    parser.add_argument("--config", type=Path, required=True)
    parser.add_argument("--out", type=Path, required=True)
    parser.add_argument("--benchmark", choices=["classification", "calibration", "all"], default="all")
    parser.add_argument(
        "--replicates",
        type=int,
        help="Override the configured replicate count when building a validation manifest.",
    )
    parser.add_argument("--shard-count", type=int, default=1)
    parser.add_argument("--shard-index", type=int, default=1)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    config = json.loads(args.config.read_text())
    benchmarks = ["classification", "calibration"] if args.benchmark == "all" else [args.benchmark]
    rows = []
    for benchmark in benchmarks:
        bench_cfg = config[benchmark]
        for cell in bench_cfg["cells"]:
            replicates = args.replicates or int(
                cell.get("replicates", bench_cfg["replicates"])
            )
            for rep in range(1, replicates + 1):
                rows.append(
                    {
                        "benchmark": benchmark,
                        "cell": cell["name"],
                        "replicate": rep,
                    }
                )
    if args.shard_count < 1 or not 1 <= args.shard_index <= args.shard_count:
        raise ValueError("shard index must be between 1 and shard count")
    rows = rows[args.shard_index - 1 :: args.shard_count]
    args.out.parent.mkdir(parents=True, exist_ok=True)
    with args.out.open("w", newline="") as handle:
        writer = csv.DictWriter(
            handle,
            fieldnames=["benchmark", "cell", "replicate"],
            delimiter="\t",
            lineterminator="\n",
        )
        writer.writeheader()
        for row in rows:
            writer.writerow(row)
    print(f"Wrote {args.out}")


if __name__ == "__main__":
    main()

File: sim/build_benchmark_manifest.R
#!/usr/bin/env Rscript

script_args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", script_args[grep(file_arg, script_args)][1]))
source(file.path(dirname(script_path), "benchmark_lib.R"))

args <- commandArgs(trailingOnly = TRUE)
get_arg <- function(flag, default = NULL) {
  idx <- match(flag, args)
  if (is.na(idx) || idx == length(args)) default else args[[idx + 1L]]
}

config_path <- get_arg("--config")
out_path <- get_arg("--out")
benchmark <- get_arg("--benchmark", "all")
if (is.null(config_path) || is.null(out_path)) {
  stop("Usage: Rscript sim/build_benchmark_manifest.R --config <json> --out <tsv> [--benchmark classification|calibration|all]", call. = FALSE)
}

config <- load_config(config_path)
benchmarks <- if (identical(benchmark, "all")) c("classification", "calibration") else benchmark
rows <- list()
ri <- 1L
for (bench in benchmarks) {
  bench_cfg <- config[[bench]]
  for (cell in bench_cfg[["cells"]]) {
    for (rep in seq_len(as.integer(bench_cfg[["replicates"]]))) {
      rows[[ri]] <- data.frame(benchmark = bench, cell = cell[["name"]], replicate = rep, stringsAsFactors = FALSE)
      ri <- ri + 1L
    }
  }
}

ensure_dir(dirname(out_path))
manifest <- do.call(rbind, rows)
write_table(out_path, manifest, c("benchmark", "cell", "replicate"), "\t")
cat(sprintf("Wrote %s\n", out_path))
